File: app/src/test/test/java/com/timeback/ui/viewmodel/AppManagementViewModelTest.java
package com.timeback.ui.viewmodel;

import com.timeback.ui.domain.model.ActionResult;
import com.timeback.ui.domain.model.AppClassification;
import com.timeback.ui.domain.model.ScreenState;
import com.timeback.ui.fake.AppManagementViewData;
import com.timeback.ui.fake.FakeFeatureGateway;
import com.timeback.ui.feature.apps.AppManagementViewModel;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * AppManagementViewModel 단위 테스트.
 */
class AppManagementViewModelTest {

    private FakeFeatureGateway gateway;
    private AppManagementViewModel viewModel;

    @BeforeEach
    void setUp() {
        gateway = new FakeFeatureGateway();
    }

    @Test
    void loadApps_shouldReturnContentWithAppList() throws Exception {
        viewModel = new AppManagementViewModel(gateway);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertNotNull(state);
        assertInstanceOf(ScreenState.Content.class, state);

        Object data = ((ScreenState.Content) state).getData();
        assertInstanceOf(AppManagementViewData.class, data);
        AppManagementViewData amd = (AppManagementViewData) data;
        assertEquals(5, amd.getApps().size());
    }

    @Test
    void loadApps_shouldContainInstagramAsWaste() throws Exception {
        viewModel = new AppManagementViewModel(gateway);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        AppManagementViewData data = (AppManagementViewData) ((ScreenState.Content) state).getData();
        assertEquals("Instagram", data.getApps().get(0).getAppName());
        assertEquals(AppClassification.WASTE, data.getApps().get(0).getClassification());
    }

    @Test
    void changeClassification_shouldReturnSuccess() throws Exception {
        viewModel = new AppManagementViewModel(gateway);
        Thread.sleep(300);

        viewModel.changeClassification("com.instagram.android", AppClassification.LEISURE);
        Thread.sleep(300);

        ActionResult result = viewModel.actionResult.getValue();
        assertNotNull(result);
        assertInstanceOf(ActionResult.Success.class, result);
    }
}

File: src/test/java/com/timeback/ui/viewmodel/AppManagementViewModelTest.kt
package com.timeback.ui.viewmodel

import com.timeback.ui.domain.model.*
import com.timeback.ui.fake.AppManagementViewData
import com.timeback.ui.fake.FakeFeatureGateway
import com.timeback.ui.feature.apps.AppManagementViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.*
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

/**
 * UI-04 AppManagementViewModel 단위 테스트
 * 검증: loadApps → CONTENT, changeClassification → SUCCESS → 재조회
 */
@OptIn(ExperimentalCoroutinesApi::class)
class AppManagementViewModelTest {

    private val testDispatcher = UnconfinedTestDispatcher()
    private lateinit var gateway: FakeFeatureGateway
    private lateinit var viewModel: AppManagementViewModel

    @BeforeEach
    fun setup() {
        Dispatchers.setMain(testDispatcher)
        gateway = FakeFeatureGateway()
        viewModel = AppManagementViewModel(gateway)
    }

    @AfterEach
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    fun `loadApps shows CONTENT with app list`() = runTest {
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
        val data = (state as ScreenState.Content<*>).data as AppManagementViewData
        assertEquals(4, data.apps.size)
    }

    @Test
    fun `changeClassification returns SUCCESS`() = runTest {
        advanceUntilIdle()

        viewModel.changeClassification("com.google.android.youtube", AppClassification.WASTE)
        advanceUntilIdle()

        val result = viewModel.actionResult.value
        assertTrue(result is ActionResult.Success)
    }
}
